mypy/expandtype.py

```diff
@@ -508,17 +508,23 @@ def expand_unpack_with_variables(
 
 
 @overload
-def expand_self_type(var: Var, typ: ProperType, replacement: ProperType) -> ProperType:
+def expand_self_type(typ: ProperType, info: TypeInfo, replacement: ProperType) -> ProperType:
     ...
 
 
 @overload
-def expand_self_type(var: Var, typ: Type, replacement: Type) -> Type:
+def expand_self_type(typ: Type, info: TypeInfo, replacement: Type) -> Type:
     ...
 
 
-def expand_self_type(var: Var, typ: Type, replacement: Type) -> Type:
+@overload
+def expand_self_type(typ: None, info: TypeInfo, replacement: Type) -> None:
+    ...
+
+
+def expand_self_type(typ: Type | None, info: TypeInfo, replacement: Type) -> Type | None:
     """Expand appearances of Self type in a variable type."""
-    if var.info.self_type is not None and not var.is_property:
-        return expand_type(typ, {var.info.self_type.id: replacement})
+    if typ is not None and info.self_type is not None:
```

Comment:
The `var.is_property` check seems to be lost here. Are you sure it's okay to not include it?

cc @ilevkivskyi who modified this code last.

Reply:
I intentionally removed it because it doesn't seem to change anything. When dealing with properties (which some tests do), it seems to function as we expect

Reply:
FWIW, properties should never need this function. It was specifically written for dealing with attribute access to something like `x: Self`. Properties should be handled by `bind_self()` (similar to regular methods), so this check would automatically guarantee that "old style" and "new style" self-types will always work the same way, without manually checking all possible corner cases.

Why do you need such dangerous things for just two-line function? You can simply write your own.

Reply:
Was trying to simplify the code a bit because they serve the same functionality, but will split it as both you and JukkaL suggested it.